### PR TITLE
Parallel scheduling

### PR DESCRIPTION
**Pull request by aluzzardi**
Work in progress for parallel scheduling for the Swarm driver.

/cc @abronan @vieux

### CODE REVIEW

File: cluster/swarm/cluster.go [hunk 1]
```diff
@@ -20,14 +20,30 @@ import (
...
+type pendingContainer struct {
+	Config *cluster.ContainerConfig
+	Name   string
+	Node   *node.Node
+}
+
+func (p *pendingContainer) ToContainer() *cluster.Container {
+	return &cluster.Container{
+		Container: dockerclient.Container{
+			Names: []string{"/" + p.Name},
```
Comment (vieux) [inline cluster/swarm/cluster.go]: should we check if name is empty here ?

**Comment by aluzzardi** [comment]
This happened. Investigating.

```
2015/10/07 02:08:49 http: panic serving 127.0.0.1:53429: runtime error: invalid memory address or nil pointer dereference
goroutine 21121 [running]:
net/http.(*conn).serve.func1(0xc82cf6bc30, 0x7f9aae2f7b30, 0xc82b21f780)
        /usr/local/go/src/net/http/server.go:1287 +0xb5
github.com/docker/swarm/cluster.Containers.Get(0xc837771c00, 0xe, 0x10, 0xc8543172c0, 0x40, 0xdbd640)
        /home/ubuntu/go/src/github.com/docker/swarm/cluster/container.go:53 +0xcd3
github.com/docker/swarm/scheduler/node.(*Node).Container(0xc834adc6e0, 0xc8543172c0, 0x40, 0x13)
        /home/ubuntu/go/src/github.com/docker/swarm/scheduler/node/node.go:47 +0x54
github.com/docker/swarm/cluster/swarm.(*Cluster).listNodes(0xc8200ae240, 0x0, 0x0, 0x0)
        /home/ubuntu/go/src/github.com/docker/swarm/cluster/swarm/cluster.go:618 +0x31b
github.com/docker/swarm/cluster/swarm.(*Cluster).createContainer(0xc8200ae240, 0xc829b0e380, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /home/ubuntu/go/src/github.com/docker/swarm/cluster/swarm/cluster.go:138 +0x3bf
github.com/docker/swarm/cluster/swarm.(*Cluster).CreateContainer(0xc8200ae240, 0xc829b0e380, 0x0, 0x0, 0x0, 0x0, 0x0)
        /home/ubuntu/go/src/github.com/docker/swarm/cluster/swarm/cluster.go:107 +0x6e
github.com/docker/swarm/api.postContainersCreate(0xc8200b72c0, 0x7f9aaa1c1aa8, 0xc829b36630, 0xc82c524d20)
        /home/ubuntu/go/src/github.com/docker/swarm/api/handlers.go:380 +0x295
github.com/docker/swarm/api.NewPrimary.func1(0x7f9aaa1c1aa8, 0xc829b36630, 0xc82c524d20)
        /home/ubuntu/go/src/github.com/docker/swarm/api/primary.go:122 +0x300
net/http.HandlerFunc.ServeHTTP(0xc820246460, 0x7f9aaa1c1aa8, 0xc829b36630, 0xc82c524d20)
        /usr/local/go/src/net/http/server.go:1422 +0x3a
github.com/gorilla/mux.(*Router).ServeHTTP(0xc8200ac140, 0x7f9aaa1c1aa8, 0xc829b36630, 0xc82c524d20)
        /home/ubuntu/go/src/github.com/docker/swarm/Godeps/_workspace/src/github.com/gorilla/mux/mux.go:98 +0x29e
github.com/docker/swarm/api.(*dispatcher).ServeHTTP(0xc8200ab040, 0x7f9aaa1c1aa8, 0xc829b36630, 0xc82c524d20)
        /home/ubuntu/go/src/github.com/docker/swarm/api/server.go:32 +0x9f
net/http.serverHandler.ServeHTTP(0xc820279860, 0x7f9aaa1c1aa8, 0xc829b36630, 0xc82c524d20)
        /usr/local/go/src/net/http/server.go:1862 +0x19e
net/http.(*conn).serve(0xc82cf6bc30)
        /usr/local/go/src/net/http/server.go:1361 +0xbee
created by net/http.(*Server).Serve
        /usr/local/go/src/net/http/server.go:1910 +0x3f6
```

**Comment by aluzzardi** [comment]
Fixed in last changes

**Comment by abronan** [comment]
I personally like the solution with the pending list :+1:, this also gets rid of the deadlock with the pull issue. Even though the lock still stays at the scheduler level (but the scheduling decision is fast enough anyway).

Will test it and tear it apart to see if there is no more panics when issuing parallel run+create+remove requests. But SGTM!

**Comment by aluzzardi** [comment]
Made a few other changes - it's good for review now.

/cc @abronan @vieux @chanwit @jimmyxian

**Comment by aluzzardi** [comment]
PTAL

**Comment by jimmyxian** [comment]
Aha, more clear than before. :+1: 
LGTM

**Comment by vieux** [comment]
LGTM, I'd like to wait for the mesos tests to get back to normal to make sure this PR doesn't break mesos if you don't mind

**Comment by abronan** [comment]
LGTM :cocktail: (when CI happy)

**Comment by chanwit** [comment]
I'm reading the commits to understand how it works. Allow me some more time.

**Comment by chanwit** [comment]
A few things there. Also LGTM when CI pass :):):)

Maybe you'd like to remove "WIP:" out of commit `a74d90ba` message :-) ?

**Comment by aluzzardi** [comment]
@chanwit Fixed! Thanks

**Pull request merged**